<a href="https://www.kaggle.com/code/ahmedfakhar123/ml-24-knn-imputer-iterative-imputer?scriptVersionId=341259931" target="_blank"><img align="left" alt="Kaggle" title="Open in Kaggle" src="https://kaggle.com/static/images/open-in-kaggle.svg"></a>

---

# 🤖 ML 24 — KNN Imputer & Iterative Imputer

### **KNN & Iterative Methods for Smart Missing Value Imputation**

---

# 📖 Introduction

Missing values are common in real-world datasets. We previously used **Simple Imputer** with mean, median, and mode, but these methods don't consider relationships between data points.

In this notebook, we'll explore two advanced techniques:

* 🔹 **KNN Imputer** — uses similar observations to estimate missing values.
* 🔹 **Iterative Imputer** — predicts missing values using other features.
* 🔹 Compare both methods with traditional imputation techniques.



---
# 📦 Imports

In [1]:
import pandas as pd
import numpy as np

from sklearn.model_selection import train_test_split

from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score

---
# 📊 Load the Dataset

We'll use the Titanic dataset and focus on a few numerical features. The target variable is `Survived`.

In [2]:
df = pd.read_csv('/kaggle/input/datasets/ahmedfakhar123/titanic-survival-dataset/Titanic_Survival_Dataset.csv')[
    ['Age','Pclass','Fare','Survived']
]
df.head()

,Age,Pclass,Fare,Survived
0,22.0,3,7.2500,0
1,38.0,1,71.2833,1
2,26.0,3,7.9250,1
3,35.0,1,53.1000,1
4,35.0,3,8.0500,0


---
# 🔍 Check Missing Values

Before applying an imputation technique, let's identify the percentage of missing values in each column.

In [3]:
df.isnull().mean() * 100

Age         19.86532
Pclass       0.00000
Fare         0.00000
Survived     0.00000
dtype: float64

---
# ✂️ Split Features and Target

In [4]:
X = df.drop(columns=['Survived'])
y = df['Survived']

Now split the data into training and testing sets.

In [5]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=2)

X_train.head()

,Age,Pclass,Fare
30,40.0,1,27.7208
10,4.0,3,16.7000
873,47.0,3,9.0000
182,9.0,3,31.3875
876,20.0,3,9.8458


> ⚠️ We split the data before imputation to avoid data leakage. The imputer learns information only from the training data.

---

# 🤖 Step 1: What is KNN Imputer?

**KNN** stands for **K-Nearest Neighbors**.

KNN Imputer fills a missing value by finding the **K most similar observations** and using their values to estimate the missing value.

For example, if a person's salary is missing, KNN can find people with similar **age, experience, and other features** and use their salaries to estimate it.


### 💡 Simple idea

**Simple Imputer:**

> "Use the column's median."

**KNN Imputer:**

> "Find similar rows and use their values."

## 📦 Import Iterative Imputer

In [6]:
from sklearn.impute import KNNImputer, SimpleImputer

## 🤖 KNN Imputation

KNN Imputer finds the nearest observations and uses them to estimate missing values.

In [7]:
KNN = KNNImputer()

X_train_KNN = KNN.fit_transform(X_train)
X_test_KNN = KNN.transform(X_test)

### Train Logistic Regression

In [8]:
model_KNN = LogisticRegression()

model_KNN.fit(X_train_KNN, y_train)

y_pred_KNN = model_KNN.predict(X_test_KNN)

### Accuracy

In [9]:
accuracy_knn = accuracy_score(y_test, y_pred_KNN)
print("Accuracy with KNNImputer:", accuracy_knn)

Accuracy with KNNImputer: 0.7039106145251397


## 📌 Simple Imputation

Now let's compare KNN Imputation with the traditional SimpleImputer.

In [10]:
simple = SimpleImputer()

X_train_simple = simple.fit_transform(X_train)
X_test_simple = simple.transform(X_test)

### Train Logistic Regression

In [11]:
model_simple = LogisticRegression()

model_simple.fit(X_train_simple, y_train)

y_pred_simple = model_simple.predict(X_test_simple)

### Accuracy

In [12]:
accuracy_simple = accuracy_score(y_test, y_pred_simple)
print("Accuracy with KNNImputer:", accuracy_simple)

Accuracy with KNNImputer: 0.6927374301675978


## 📊 Comparing the Results

In [13]:
print(f"KNN Imputer:    {accuracy_knn:.4f}")
print(f"Simple Imputer: {accuracy_simple:.4f}")

KNN Imputer:    0.7039
Simple Imputer: 0.6927


In this experiment, **KNN Imputer performs slightly better** than Simple Imputer.

| Imputation Method |   Accuracy |
| ----------------- | ---------: |
| Simple Imputer    |     0.6983 |
| KNN Imputer       | **0.7039** |

This doesn't mean KNN will always perform better. The best imputation method depends on the **dataset, features, missing-value pattern, and ML model**.


---
# 🔄 Step 2: What is Iterative Imputer?

**Iterative Imputer** estimates missing values by modeling each feature with missing values as a function of the other features.

Unlike `SimpleImputer`, which uses a fixed statistic, Iterative Imputer **learns relationships between features** to predict the missing values.

## 📦 Import Iterative Imputer

`IterativeImputer` is experimental in Scikit-learn, so we need to enable it first:


In [14]:
from sklearn.experimental import enable_iterative_imputer
from sklearn.impute import IterativeImputer

## 🛠️ Apply Iterative Imputation

In [15]:
iterative_imputer = IterativeImputer(random_state=2)

X_train_iterative = iterative_imputer.fit_transform(X_train)
X_test_iterative = iterative_imputer.transform(X_test)

## 🤖 Train Logistic Regression

In [16]:
model_iterative = LogisticRegression()

model_iterative.fit(X_train_iterative, y_train)

y_pred_iterative = model_iterative.predict(X_test_iterative)

accuracy_iterative = accuracy_score(y_test, y_pred_iterative)

print(f"Accuracy with Iterative Imputer: {accuracy_iterative:.4f}")

Accuracy with Iterative Imputer: 0.6927


## 💡 How it works

For example, if `Age` has missing values, Iterative Imputer can use features like **Pclass** and **Fare** to predict those missing `Age` values.

**Simple Imputer:**

> Missing Age → replace with median Age

**KNN Imputer:**

> Missing Age → find similar observations

**Iterative Imputer:**

> Missing Age → use other features to predict Age

This makes Iterative Imputation particularly useful when **features have meaningful relationships with each other**.


---
# 🏁 Conclusion

In this notebook, we explored advanced techniques for handling **missing values** using Scikit-learn.

* 🔹 **KNN Imputer** fills missing values using information from the most similar observations.
* 🔹 **Iterative Imputer** predicts missing values by learning relationships between different features.
* 🔹 We compared these techniques with the traditional **Simple Imputer**.
* 🔹 We also learned the importance of fitting the imputer only on the **training data** to avoid data leakage.

Overall, KNN and Iterative Imputation provide more intelligent alternatives to simple mean/median imputation, especially when relationships between features can help estimate missing values.

---